pynicotine/portmapper.py

```diff
@@ -126,17 +127,23 @@ def _request_port_mapping(self, public_port, private_port, lease_duration):
 
         with socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM, proto=socket.IPPROTO_UDP) as sock:
             sock.bind((self.local_ip_address, 0))
-            sock.settimeout(self.REQUEST_TIMEOUT)
-
             request = self.PortmapRequest(public_port, private_port, lease_duration)
-            request.sendto(sock, (self._gateway_address, self.REQUEST_PORT))
+            timeout = self.REQUEST_INIT_TIMEOUT
+            i = 1
 
-            try:
-                response = self.PortmapResponse(message=sock.recv(16))
-                return response.result
+            while i <= self.REQUEST_ATTEMPTS:
```

Comment:
Suggest to avoid any risk of an infinite loop, do away with `i`:

`for num_attempt in range(1, self.REQUEST_ATTEMPTS):`